src/Microsoft.ML.Dnn/ImageClassificationTransform.cs

```diff
@@ -445,6 +451,24 @@ private IDataView GetShuffledData(string path)
                         statisticsCallback(metrics);
                     }
                 }
+                // early stopping check
+                if (options.EnableEarlyStopping)
+                {
+                    if (metrics.Train.Accuracy - options.MinDelta > history.Accuracy)
+                    {
+                        history.Accuracy = metrics.Train.Accuracy;
+                        wait = 0;
+                    }
```

Comment:
This is incorrect. It needs to be absolute change as documented here: [https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/keras/callbacks.py#L1143-L1260](https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/keras/callbacks.py#L1143-L1260)

min_delta: Minimum change in the monitored quantity
          to qualify as an improvement, i.e. an absolute
          change of less than min_delta, will count as no
          improvement.

Looking at the code:

 if mode == 'min':
      self.monitor_op = np.less
    elif mode == 'max':
      self.monitor_op = np.greater
    else:
      if 'acc' in self.monitor:
        self.monitor_op = np.greater
      else:
        self.monitor_op = np.less

    if self.monitor_op == np.greater:
      self.min_delta *= 1
    else:
      self.min_delta *= -1

The last 4 lines change the sign of the delta and that takes care of absolute difference in below function:

def on_epoch_end(self, epoch, logs=None):
    current = self.get_monitor_value(logs)
    if current is None:
      return
    if self.monitor_op(current - self.min_delta, self.best):
      self.best = current
      self.wait = 0
      if self.restore_best_weights:
        self.best_weights = self.model.get_weights()
    else:
      self.wait += 1
      if self.wait >= self.patience:
        self.stopped_epoch = epoch
        self.model.stop_training = True
        if self.restore_best_weights:
          if self.verbose > 0:
            print('Restoring model weights from the end of the best epoch.')
          self.model.set_weights(self.best_weights)

Reply:
As we discussed offline, the change in sign of the min_delta doesn't handle taking the absolute value. Even though it is mentioned in the comment that they take absolute value of the change, it is not implemented that way in the code. However, we want to consider absolute value and I will make that change.

Reply:
As we discussed offline, the change in sign of the min_delta doesn't handle taking the absolute value. Even though it is mentioned in the comment that they take absolute value of the change, it is not implemented that way in the code. However, we want to consider absolute value and I will make that change.

---
In reply to: [328248944](https://github.com/dotnet/machinelearning/pull/4237#discussion_r328248944) [](ancestors = 328248944)